src/v/pandaproxy/kafka_client_cache.h

```diff
@@ -60,6 +61,14 @@ class kafka_client_cache {
     size_t size() const;
     size_t max_size() const;
 
+    template<std::invocable<kafka_client_cache&> Func>
+    auto invoke_on_cache(Func&& func) {
+        return ss::try_with_gate(_gc_gate, [this, &func] {
+            return _gc_lock.with(
+              [this, &func] { return std::forward<Func>(func)(*this); });
```

Comment:
Here too:
```
return _gc_lock.with([this, func = std::forward<Func>(func)] { return func(*this); } );
```

Reply:
another option would be to always move:
```
template<std::invocable<kafka_client_cache&> Func>
auto invoke_on_cache(Func func) {
  .... _gc_gate, [this, func = std::move(func)] ....
```
 or use coroutines:
 ```
template<std::invocable<kafka_client_cache&> Func>
auto invoke_on_cache(Func func) {
   auto guard = gate_guard(_gc_gate);
   auto units =  _gc_lock.get_units();
   co_await func(*this);
}
```

Reply:
coroutines are not an option because some of the methods that call `invoke_on_cache` do not return a promise_type.

I'm running into several compile errors when I forward the func in the capture lists so I'll get to this comment after I resolve everything else.

Reply:
This is going to take more time to get working:
- The compiler doesn't like Moving/Forwarding the func in the capture list because the copy-constructor is automatically called and the request_t object in server.h has a unique_ptr. So I'd have to refactor many places to remove that unique_ptr
- The compiler doesn't like coroutines here because several methods that call invoke_on_cache do not return a promise_type

Reply:
1. Moving shouldn't invoke the copy constructor.
2. It would probably be easier to `ss::futurize` the calls.

Reply:
Finally got it working. Apparently I forgot the cardinal rule that lambda must be `mutable` if you capture by move.